File: Sources/ClaudexCore/Presentation/Presentation.swift
import Foundation

// MARK: - Lamps

public enum LampMode: Sendable, Equatable {
    case off, dim, steady, breathing, blinking
}

/// The three lamps of one traffic light.
public struct LampTriple: Sendable, Equatable {
    public var red: LampMode
    public var yellow: LampMode
    public var green: LampMode

    public init(red: LampMode = .off, yellow: LampMode = .off, green: LampMode = .off) {
        self.red = red
        self.yellow = yellow
        self.green = green
    }

    public static let allOff = LampTriple()

    public init(_ lamps: AgentLamps) {
        self.init(
            red: lamps.red ? .blinking : .off,
            yellow: lamps.yellow == .fresh ? .steady : (lamps.yellow == .parked ? .dim : .off),
            green: lamps.green ? .breathing : .off
        )
    }

    /// The single lamp shown next to one session.
    public static func single(_ state: SessionState, fresh: Bool) -> LampTriple {
        switch state {
        case .needsAttention: return LampTriple(red: .blinking)
        case .waitingForUser: return LampTriple(yellow: fresh ? .steady : .dim)
        case .working: return LampTriple(green: .breathing)
        }
    }

    public var isAllOff: Bool { red == .off && yellow == .off && green == .off }
}

// MARK: - Formatting

public enum Formatters {
    /// "now", "45s", "12m", "3h", "2d".
    public static func elapsed(since: Date, now: Date) -> String {
        let s = max(0, now.timeIntervalSince(since))
        if s < 5 { return "now" }
        return Durations.short(s)
    }

    /// "2h 13m", "45m", "3d 4h".
    public static func countdown(_ interval: TimeInterval) -> String {
        let s = max(0, Int(interval))
        if s < 60 { return "<1m" }
        let d = s / 86_400, h = (s % 86_400) / 3600, m = (s % 3600) / 60
        if d > 0 { return h > 0 ? "\(d)d \(h)h" : "\(d)d" }
        if h > 0 { return m > 0 ? "\(h)h \(m)m" : "\(h)h" }
        return "\(m)m"
    }

    /// "resets in 2h 13m", "≈ resets in 40m", "reset", or nil when unknown.
    public static func reset(_ window: UsageWindow, now: Date) -> String? {
        if window.isReset { return "reset" }
        guard let r = window.resetsAt, r > now else { return nil }
        return (window.resetIsEstimate ? "≈ " : "") + "resets in " + countdown(r.timeIntervalSince(now))
    }

    /// Compact form for tight columns: "↻ 2h 13m", "≈↻ 40m", "reset".
    public static func shortReset(_ window: UsageWindow, now: Date) -> String? {
        if window.isReset { return "reset" }
        guard let r = window.resetsAt, r > now else { return nil }
        return (window.resetIsEstimate ? "≈" : "") + "↻ " + countdown(r.timeIntervalSince(now))
    }

    public static func percent(_ v: Double?) -> String {
        guard let v else { return "—" }
        return "\(Int(v.rounded()))%"
    }
}

public enum UsageLevel: Sendable, Equatable {
    case normal, elevated, high, critical

    public init(percent: Double?) {
        let p = percent ?? 0
        if p >= 100 { self = .critical } else if p >= 80 { self = .high } else if p >= 50 { self = .elevated } else { self = .normal }
    }
}

// MARK: - Peek events

public struct PeekEvent: Sendable, Equatable, Identifiable {
    public enum Kind: Sendable, Equatable {
        case attention, usage, finished

        public var priority: Int {
            switch self {
            case .attention: return 3
            case .usage: return 2
            case .finished: return 1
            }
        }

        public var duration: TimeInterval {
            switch self {
            case .attention: return 5
            case .usage: return 4
            case .finished: return 3
            }
        }
    }

    public var id: UUID
    /// Coalescing key, e.g. `attn:<session>`, `done:<session>`, `usage:<window>`.
    public var key: String
    public var kind: Kind
    public var provider: Provider
    public var lamp: LampColor
    public var title: String
    public var detail: String
    public var sessionID: String?
    public var createdAt: Date

    public init(key: String, kind: Kind, provider: Provider, lamp: LampColor, title: String, detail: String,
                sessionID: String?, createdAt: Date, id: UUID = UUID()) {
        self.id = id
        self.key = key
        self.kind = kind
        self.provider = provider
        self.lamp = lamp
        self.title = title
        self.detail = detail
        self.sessionID = sessionID
        self.createdAt = createdAt
    }
}

/// Turns snapshot changes into peek events. The first snapshot after launch never peeks.
public struct PeekEventDetector: Sendable {
    public var finishedMinimumTurn: TimeInterval = 20
    public var thresholds: [Double] = [80, 100]
    private var previous: [String: AgentSession] = [:]
    private var sessionsInitialized = false
    private var usageSeen: [String: Double] = [:]
    private var usageInitialized = false

    public init() {}

    public mutating func diff(sessions: [AgentSession], now: Date) -> [PeekEvent] {
        defer {
            previous = Dictionary(uniqueKeysWithValues: sessions.map { ($0.id, $0) })
            sessionsInitialized = true
        }
        guard sessionsInitialized else { return [] }
        var events: [PeekEvent] = []
        for s in sessions {
            let prev = previous[s.id]
            let name = "\(s.provider.displayName) · \(s.project ?? s.title)"
            switch s.state {
            case let .needsAttention(reason):
                if prev?.state != s.state {
                    events.append(PeekEvent(key: "attn:\(s.id)", kind: .attention, provider: s.provider, lamp: .red,
                                            title: name, detail: Self.attentionDetail(reason, activity: s.activity),
                                            sessionID: s.id, createdAt: now))
                }
            case .waitingForUser:
                if let prev, prev.state == .working, now.timeIntervalSince(prev.stateSince) >= finishedMinimumTurn {
                    events.append(PeekEvent(key: "done:\(s.id)", kind: .finished, provider: s.provider, lamp: .yellow,
                                            title: "Done · \(s.project ?? s.title)", detail: s.provider.displayName,
                                            sessionID: s.id, createdAt: now))
                }
            case .working:
                break
            }
        }
        return events
    }

    static func attentionDetail(_ reason: AttentionReason, activity: String) -> String {
        switch reason {
        case let .approval(tool, _): return "needs approval" + (tool.map { " — \($0)" } ?? "")
        case let .question(h): return "has a question" + (h.map { " — \($0)" } ?? "")
        case .input: return activity
        case .error: return activity
        case .usageLimit: return "hit the usage limit"
        }
    }

    public mutating func diff(usage: UsageSnapshot, now: Date) -> [PeekEvent] {
        var events: [PeekEvent] = []
        for provider in Provider.allCases {
            for w in usage[provider].windows {
                guard let v = w.usedPercent else { continue }
                let key = "\(provider.rawValue):\(w.id)"
                let old = usageSeen[key]
                usageSeen[key] = v
                guard usageInitialized, let old else { continue }
                for t in thresholds where old < t && v >= t {
                    let reset = Formatters.reset(w, now: now).map { " · \($0)" } ?? ""
                    events.append(PeekEvent(key: "usage:\(key)", kind: .usage, provider: provider,
                                            lamp: t >= 100 ? .red : .yellow,
                                            title: "\(provider.displayName) \(w.label) at \(Int(t))%",
                                            detail: (t >= 100 ? "limit reached" : "usage is high") + reset,
                                            sessionID: nil, createdAt: now))
                }
            }
        }
        if !usage.claude.windows.isEmpty || !usage.codex.windows.isEmpty { usageInitialized = true }
        return events
    }
}

/// Priority queue of peeks with coalescing, preemption and hover-hold.
public struct PeekQueue: Sendable {
    public enum Effect: Sendable, Equatable {
        case none
        case show(PeekEvent)
        case hide
    }

    public private(set) var current: PeekEvent?
    public private(set) var shownAt: Date?
    public private(set) var deadline: Date?
    public private(set) var queued: [PeekEvent] = []
    public private(set) var isHeld = false
    private var heldRemaining: TimeInterval?

    public var maxQueued = 5
    public var maxQueueAge: TimeInterval = 30
    public var minimumShowBeforePreempt: TimeInterval = 1.2

    public init() {}

    public mutating func enqueue(_ e: PeekEvent, now: Date) -> Effect {
        if var cur = current, cur.key == e.key {
            cur.title = e.title
            cur.detail = e.detail
            current = cur
            if !isHeld { deadline = max(deadline ?? now, now.addingTimeInterval(2)) }
            return .show(cur)
        }
        queued.removeAll { $0.key == e.key }
        guard let cur = current else {
            return show(e, now: now)
        }
        let canPreempt = !isHeld && now.timeIntervalSince(shownAt ?? now) >= minimumShowBeforePreempt
        if e.kind.priority > cur.kind.priority && canPreempt {
            if cur.kind != .finished { insert(cur) }
            return show(e, now: now)
        }
        insert(e)
        return .none
    }

    /// The current peek's timer fired.
    public mutating func expire(now: Date) -> Effect {
        guard !isHeld, let d = deadline, now >= d.addingTimeInterval(-0.05) else { return .none }
        current = nil
        shownAt = nil
        deadline = nil
        while let next = queued.first {
            queued.removeFirst()
            if now.timeIntervalSince(next.createdAt) <= maxQueueAge { return show(next, now: now) }
        }
        return .hide
    }

    public mutating func hold(now: Date) {
        guard !isHeld, current != nil else { return }
        isHeld = true
        heldRemaining = deadline.map { max(0, $0.timeIntervalSince(now)) }
        deadline = nil
    }

    public mutating func release(now: Date) {
        guard isHeld else { return }
        isHeld = false
        if current != nil { deadline = now.addingTimeInterval(max(1.5, heldRemaining ?? 1.5)) }
        heldRemaining = nil
    }

    public mutating func clear() {
        current = nil
        shownAt = nil
        deadline = nil
        queued.removeAll()
        isHeld = false
        heldRemaining = nil
    }

    private mutating func show(_ e: PeekEvent, now: Date) -> Effect {
        current = e
        shownAt = now
        deadline = now.addingTimeInterval(e.kind.duration)
        return .show(e)
    }

    private mutating func insert(_ e: PeekEvent) {
        queued.append(e)
        queued.sort { a, b in
            a.kind.priority != b.kind.priority ? a.kind.priority > b.kind.priority : a.createdAt < b.createdAt
        }
        if queued.count > maxQueued { queued.removeLast(queued.count - maxQueued) }
    }
}

File: windows/src/ClaudexBar.Core/Presentation.cs
namespace ClaudexBar.Core;

public enum LampMode { Off, Dim, Steady, Breathing, Blinking }

public readonly record struct LampTriple(LampMode Red, LampMode Yellow, LampMode Green)
{
    public static readonly LampTriple AllOff = new(LampMode.Off, LampMode.Off, LampMode.Off);

    public static LampTriple From(AgentLamps l) => new(
        l.Red ? LampMode.Blinking : LampMode.Off,
        l.Yellow == YellowLamp.Fresh ? LampMode.Steady : l.Yellow == YellowLamp.Parked ? LampMode.Dim : LampMode.Off,
        l.Green ? LampMode.Breathing : LampMode.Off);

    public static LampTriple Single(SessionState state, bool fresh) => state.Kind switch
    {
        StateKind.NeedsAttention => new(LampMode.Blinking, LampMode.Off, LampMode.Off),
        StateKind.WaitingForUser => new(LampMode.Off, fresh ? LampMode.Steady : LampMode.Dim, LampMode.Off),
        _ => new(LampMode.Off, LampMode.Off, LampMode.Breathing),
    };

    public bool IsAllOff => Red == LampMode.Off && Yellow == LampMode.Off && Green == LampMode.Off;
}

public enum UsageLevel { Normal, Elevated, High, Critical }

public static class Formatters
{
    public static UsageLevel Level(double? percent) =>
        (percent ?? 0) >= 100 ? UsageLevel.Critical : (percent ?? 0) >= 80 ? UsageLevel.High
        : (percent ?? 0) >= 50 ? UsageLevel.Elevated : UsageLevel.Normal;

    public static string Elapsed(DateTimeOffset since, DateTimeOffset now) =>
        now - since < TimeSpan.FromSeconds(5) ? "now" : Durations.Short(now - since);

    public static string Countdown(TimeSpan interval)
    {
        long s = Math.Max(0, (long)interval.TotalSeconds);
        if (s < 60) return "<1m";
        long d = s / 86_400, h = s % 86_400 / 3600, m = s % 3600 / 60;
        if (d > 0) return h > 0 ? $"{d}d {h}h" : $"{d}d";
        if (h > 0) return m > 0 ? $"{h}h {m}m" : $"{h}h";
        return $"{m}m";
    }

    public static string? Reset(UsageWindow w, DateTimeOffset now)
    {
        if (w.IsReset) return "reset";
        if (w.ResetsAt is not { } r || r <= now) return null;
        return (w.ResetIsEstimate ? "≈ " : "") + "resets in " + Countdown(r - now);
    }

    public static string? ShortReset(UsageWindow w, DateTimeOffset now)
    {
        if (w.IsReset) return "reset";
        if (w.ResetsAt is not { } r || r <= now) return null;
        return (w.ResetIsEstimate ? "≈" : "") + "↻ " + Countdown(r - now);
    }

    public static string Percent(double? v) => v is { } p ? $"{(int)Math.Round(p)}%" : "—";
}

public enum PeekKind { Attention = 3, Usage = 2, Finished = 1 }

public sealed record PeekEvent(string Key, PeekKind Kind, Provider Provider, LampColor Lamp, string Title, string Detail,
    string? SessionId, DateTimeOffset CreatedAt)
{
    public Guid Id { get; init; } = Guid.NewGuid();
    public int Priority => (int)Kind;
    public TimeSpan Duration => Kind switch
    {
        PeekKind.Attention => TimeSpan.FromSeconds(5),
        PeekKind.Usage => TimeSpan.FromSeconds(4),
        _ => TimeSpan.FromSeconds(3),
    };
}

/// <summary>Turns snapshot changes into peek events. The first snapshot never peeks.</summary>
public sealed class PeekEventDetector
{
    public TimeSpan FinishedMinimumTurn { get; init; } = TimeSpan.FromSeconds(20);
    private Dictionary<string, AgentSession> _previous = [];
    private bool _sessionsInitialized;
    private readonly Dictionary<string, double> _usageSeen = [];
    private bool _usageInitialized;

    public List<PeekEvent> Diff(IReadOnlyList<AgentSession> sessions, DateTimeOffset now)
    {
        var events = new List<PeekEvent>();
        if (_sessionsInitialized)
        {
            foreach (var s in sessions)
            {
                _previous.TryGetValue(s.Id, out var prev);
                var name = $"{s.Provider.DisplayName()} · {s.Project ?? s.Title}";
                if (s.State.Kind == StateKind.NeedsAttention && prev?.State != s.State)
                    events.Add(new PeekEvent($"attn:{s.Id}", PeekKind.Attention, s.Provider, LampColor.Red, name,
                        AttentionDetail(s.State.Reason!, s.Activity), s.Id, now));
                else if (s.State.Kind == StateKind.WaitingForUser && prev is { State.Kind: StateKind.Working }
                         && now - prev.StateSince >= FinishedMinimumTurn)
                    events.Add(new PeekEvent($"done:{s.Id}", PeekKind.Finished, s.Provider, LampColor.Yellow,
                        $"Done · {s.Project ?? s.Title}", s.Provider.DisplayName(), s.Id, now));
            }
        }
        _previous = sessions.ToDictionary(s => s.Id);
        _sessionsInitialized = true;
        return events;
    }

    private static string AttentionDetail(AttentionReason r, string activity) => r.Kind switch
    {
        AttentionKind.Approval => "needs approval" + (r.Tool is null ? "" : $" — {r.Tool}"),
        AttentionKind.Question => "has a question" + (r.Detail is null ? "" : $" — {r.Detail}"),
        AttentionKind.UsageLimit => "hit the usage limit",
        _ => activity,
    };

    public List<PeekEvent> Diff(UsageSnapshot usage, DateTimeOffset now)
    {
        var events = new List<PeekEvent>();
        foreach (var p in ProviderInfo.All)
            foreach (var w in usage[p].Windows)
            {
                if (w.UsedPercent is not { } v) continue;
                var key = $"{p}:{w.Id}";
                bool had = _usageSeen.TryGetValue(key, out var old);
                _usageSeen[key] = v;
                if (!_usageInitialized || !had) continue;
                foreach (var t in new double[] { 80, 100 })
                {
                    if (!(old < t && v >= t)) continue;
                    var reset = Formatters.Reset(w, now) is { } r ? $" · {r}" : "";
                    events.Add(new PeekEvent($"usage:{key}", PeekKind.Usage, p, t >= 100 ? LampColor.Red : LampColor.Yellow,
                        $"{p.DisplayName()} {w.Label} at {(int)t}%", (t >= 100 ? "limit reached" : "usage is high") + reset, null, now));
                }
            }
        if (usage.Claude.Windows.Count > 0 || usage.Codex.Windows.Count > 0) _usageInitialized = true;
        return events;
    }
}

/// <summary>Priority queue of peeks with coalescing, preemption and hover-hold.</summary>
public sealed class PeekQueue
{
    public enum EffectKind { None, Show, Hide }

    public readonly record struct Effect(EffectKind Kind, PeekEvent? Event = null)
    {
        public static readonly Effect Nothing = new(EffectKind.None);
        public static readonly Effect HideIsland = new(EffectKind.Hide);
    }

    public PeekEvent? Current { get; private set; }
    public DateTimeOffset? ShownAt { get; private set; }
    public DateTimeOffset? Deadline { get; private set; }
    public List<PeekEvent> Queued { get; } = [];
    public bool IsHeld { get; private set; }
    private TimeSpan? _heldRemaining;

    public Effect Enqueue(PeekEvent e, DateTimeOffset now)
    {
        if (Current is { } cur && cur.Key == e.Key)
        {
            Current = cur with { Title = e.Title, Detail = e.Detail };
            if (!IsHeld) Deadline = Max(Deadline ?? now, now.AddSeconds(2));
            return new Effect(EffectKind.Show, Current);
        }
        Queued.RemoveAll(q => q.Key == e.Key);
        if (Current is null) return Show(e, now);
        bool canPreempt = !IsHeld && now - (ShownAt ?? now) >= TimeSpan.FromSeconds(1.2);
        if (e.Priority > Current.Priority && canPreempt)
        {
            if (Current.Kind != PeekKind.Finished) Insert(Current);
            return Show(e, now);
        }
        Insert(e);
        return Effect.Nothing;
    }

    public Effect Expire(DateTimeOffset now)
    {
        if (IsHeld || Deadline is not { } d || now < d.AddMilliseconds(-50)) return Effect.Nothing;
        Current = null;
        ShownAt = null;
        Deadline = null;
        while (Queued.Count > 0)
        {
            var next = Queued[0];
            Queued.RemoveAt(0);
            if (now - next.CreatedAt <= TimeSpan.FromSeconds(30)) return Show(next, now);
        }
        return Effect.HideIsland;
    }

    public void Hold(DateTimeOffset now)
    {
        if (IsHeld || Current is null) return;
        IsHeld = true;
        _heldRemaining = Deadline is { } d ? Max(d, now) - now : null;
        Deadline = null;
    }

    public void Release(DateTimeOffset now)
    {
        if (!IsHeld) return;
        IsHeld = false;
        if (Current is not null) Deadline = now + TimeSpan.FromSeconds(Math.Max(1.5, _heldRemaining?.TotalSeconds ?? 1.5));
        _heldRemaining = null;
    }

    public void Clear()
    {
        Current = null;
        ShownAt = null;
        Deadline = null;
        Queued.Clear();
        IsHeld = false;
        _heldRemaining = null;
    }

    private Effect Show(PeekEvent e, DateTimeOffset now)
    {
        Current = e;
        ShownAt = now;
        Deadline = now + e.Duration;
        return new Effect(EffectKind.Show, e);
    }

    private void Insert(PeekEvent e)
    {
        Queued.Add(e);
        Queued.Sort((a, b) => a.Priority != b.Priority ? b.Priority.CompareTo(a.Priority) : a.CreatedAt.CompareTo(b.CreatedAt));
        if (Queued.Count > 5) Queued.RemoveRange(5, Queued.Count - 5);
    }

    private static DateTimeOffset Max(DateTimeOffset a, DateTimeOffset b) => a > b ? a : b;
}
